### ISSUE TITLE
Improving test coverage

### ISSUE THREAD

**Issue by benanne** [Lasagne/Lasagne#112]
As we discussed in #67, we want to encourage contributors to add tests to their pull requests as much as possible. But as long as the library has poor coverage, it's going to be hard to do this. So we should probably make a concentrated effort to get the coverage to an appropriate level.

@dnouri already did some great work setting things up and adding a ton of tests (see #40 and #47 among others).

Here's the current coverage report that I get when I run py.test:

```
Name                                 Stmts   Miss  Cover   Missing
------------------------------------------------------------------
lasagne/__init__                         7      0   100%   
lasagne/init                            63      3    95%   17, 44, 96
lasagne/layers/__init__                 12      0   100%   
lasagne/layers/base                     52      5    90%   163, 211, 216, 262, 289
lasagne/layers/conv                    116     94    19%   22-42, 45-46, 49, 52, 55-64, 69-92, 99-119, 122-123, 126, 129, 132-144, 149-173
lasagne/layers/corrmm                   69     51    26%   33-72, 75-76, 79, 82, 85-89, 92-107
lasagne/layers/cuda_convnet            175    140    20%   38-93, 96-101, 104, 107, 110-123, 126-151, 156-177, 180-195, 198-207, 222, 225, 236, 239, 254-273, 276, 279, 282, 285-297
lasagne/layers/dense                    66     29    56%   101-120, 123, 126, 129, 132-146
lasagne/layers/dnn                      96     70    27%   16, 32-35, 38-41, 44-46, 51, 57-103, 106-107, 110, 113, 116-120, 123-144
lasagne/layers/helper                   45      1    98%   58
lasagne/layers/input                    24      2    92%   42-43
lasagne/layers/merge                    39      9    77%   23-26, 63-66, 70-72
lasagne/layers/noise                    31      0   100%   
lasagne/layers/pool                     79     58    27%   20-22, 25-34, 37, 58-65, 69-71, 74-85, 102-108, 112-134, 143-144, 147, 150
lasagne/layers/shape                    28     13    54%   21, 24, 31-34, 37-44, 47
lasagne/nonlinearities                  10      0   100%   
lasagne/objectives                      15      0   100%   
lasagne/regularization                   8      4    50%   8-13
lasagne/tests/conftest                   6      0   100%   
lasagne/tests/layers/conftest            6      0   100%   
lasagne/tests/layers/test_base          90      0   100%   
lasagne/tests/layers/test_dense         55      0   100%   
lasagne/tests/layers/test_input         23      0   100%   
lasagne/tests/layers/test_merge         26      0   100%   
lasagne/tests/layers/test_noise         55      0   100%   
lasagne/tests/test_examples             28     13    54%   22-23, 29-41
lasagne/tests/test_init                 64      0   100%   
lasagne/tests/test_nonlinearities        6      0   100%   
lasagne/tests/test_objectives           36      0   100%   
lasagne/theano_extensions/__init__       0      0   100%   
lasagne/theano_extensions/conv         113    104     8%   17-35, 42-56, 63-77, 86-116, 123-170, 177-211
lasagne/theano_extensions/padding       20     16    20%   16-38
lasagne/updates                         59     50    15%   12-18, 22-31, 37-47, 55-64, 75-84, 96-114
lasagne/utils                           45      9    80%   20-21, 35-36, 48-51, 98, 108
------------------------------------------------------------------
TOTAL                                 1567    671    57%   
```

It looks like the main offenders are the various convolution / pooling implementations, so I guess that's mostly going to fall to me. Hopefully I can find some time for it soon. If anyone else can contribute, please do. I think this and documentation are the two most important things we should focus on right now (we can't make a proper release without them).

Maybe we should make a checklist of missing tests in this issue that we can check off as they are added.

**Comment by dnouri** [comment]
Have you tried with --runslow?

Sander Dieleman wrote:
> It looks like the main offenders are the various convolution / pooling
> implementations, so I guess that's mostly going to fall to me. Hopefully
> I can find some time for it soon. If anyone else can contribute, please
> do. I think this and documentation are the two most important things we
> should focus on right now (we can't make a proper release without them).

I pondered a bit about writing tests for those layers and it seemed
hard.  I thought maybe the best way would be to test with real (minimal)
examples.  Do the math for the convolution by hand, send in the input
and check that the output is what you expect.  And that's a lot of work
already, given all the different cases that'd need to be tested.  But I
don't think mocking will have a lot of value here.  We *could* check
that arguments are passed on correctly into Theano, but who knows if
they're the right ones, or there's not a bug in the interaction.

**Comment by benanne** [comment]
I guess we should test output correctness at least for the 1D convolution implementations in `theano_extensions.conv`. For the layers themselves I agree it might not be essential (that's Theano's job), but it would still be good to test all the various nuts and bolts (the `dimshuffle` argument of the cuda-convnet based layers for example). So I'll try to cover those things at least.

**Comment by benanne** [comment]
Here's the output with --runslow

```
Name                                 Stmts   Miss  Cover   Missing
------------------------------------------------------------------
lasagne/__init__                         7      0   100%   
lasagne/init                            63      3    95%   17, 44, 96
lasagne/layers/__init__                 12      0   100%   
lasagne/layers/base                     52      5    90%   163, 211, 216, 262, 289
lasagne/layers/conv                    116     64    45%   22-42, 45-46, 49, 52, 55-64, 69-92, 101, 114, 116-117, 135-142, 157-164, 167, 169
lasagne/layers/corrmm                   69     51    26%   33-72, 75-76, 79, 82, 85-89, 92-107
lasagne/layers/cuda_convnet            175     68    61%   42, 47, 50, 53, 64, 68-79, 83, 85-89, 100-101, 114-115, 123, 130, 133, 141, 151, 160, 163, 168, 174, 185-187, 195, 207, 222, 225, 236, 239, 254-273, 276, 279, 282, 285-297
lasagne/layers/dense                    66     29    56%   101-120, 123, 126, 129, 132-146
lasagne/layers/dnn                      96     35    64%   16, 44-46, 59, 66, 72, 77-94, 98, 100-101, 110, 113, 126-144
lasagne/layers/helper                   45      1    98%   58
lasagne/layers/input                    24      2    92%   42-43
lasagne/layers/merge                    39      9    77%   23-26, 63-66, 70-72
lasagne/layers/noise                    31      0   100%   
lasagne/layers/pool                     79     49    38%   28-29, 58-65, 69-71, 74-85, 102-108, 112-134, 143-144, 147, 150
lasagne/layers/shape                    28     13    54%   21, 24, 31-34, 37-44, 47
lasagne/nonlinearities                  10      0   100%   
lasagne/objectives                      15      0   100%   
lasagne/regularization                   8      4    50%   8-13
lasagne/tests/conftest                   6      1    83%   10
lasagne/tests/layers/conftest            6      0   100%   
lasagne/tests/layers/test_base          90      0   100%   
lasagne/tests/layers/test_dense         55      0   100%   
lasagne/tests/layers/test_input         23      0   100%   
lasagne/tests/layers/test_merge         26      0   100%   
lasagne/tests/layers/test_noise         55      0   100%   
lasagne/tests/test_examples             28      4    86%   33-36, 41
lasagne/tests/test_init                 64      0   100%   
lasagne/tests/test_nonlinearities        6      0   100%   
lasagne/tests/test_objectives           36      0   100%   
lasagne/theano_extensions/__init__       0      0   100%   
lasagne/theano_extensions/conv         113    104     8%   17-35, 42-56, 63-77, 86-116, 123-170, 177-211
lasagne/theano_extensions/padding       20     16    20%   16-38
lasagne/updates                         59     41    31%   12-18, 22-31, 55-64, 75-84, 96-114
lasagne/utils                           45      9    80%   20-21, 35-36, 48-51, 98, 108
------------------------------------------------------------------
TOTAL                                 1567    508    68%   
```

**Comment by dimatura** [comment]
Regarding convolution, one area where I can see bugs happening is on calculations for output size / padding / border / strides. A while ago while I was still using cuda-convnet layers there was definitely a bug where `get_output_size` gave a different answer than the size of what was actually computed, but I didn't investigate enough to find out if this was a problem with theano or with lasagne.

**Comment by benanne** [comment]
Good point, we should definitely test `get_output_shape` thoroughly for those layers.

Time for an update now that #114 is merged:

```
Name                                 Stmts   Miss  Cover   Missing
------------------------------------------------------------------
lasagne/__init__                         7      0   100%   
lasagne/conftest                         3      0   100%   
lasagne/init                            63      3    95%   17, 44, 96
lasagne/layers/__init__                 21      4    81%   13-14, 18-19
lasagne/layers/base                     52      5    90%   163, 211, 216, 262, 289
lasagne/layers/conv                    116     61    47%   22-42, 45-46, 49, 52, 55-64, 69-92, 101, 114, 135-142, 157-164, 167
lasagne/layers/corrmm                   71     18    75%   25, 39, 50, 54-65, 69, 74, 83, 86, 105, 109
lasagne/layers/cuda_convnet            177     64    64%   27, 45, 50, 53, 56, 67, 71-82, 86, 92, 103-104, 117-118, 126, 133, 154, 163, 166, 171, 177, 188-190, 198, 210, 225, 228, 239, 242, 257-276, 279, 282, 285, 288-300
lasagne/layers/dense                    66     29    56%   101-120, 123, 126, 129, 132-146
lasagne/layers/dnn                      90     81    10%   16-137
lasagne/layers/helper                   45      1    98%   58
lasagne/layers/input                    24      2    92%   42-43
lasagne/layers/merge                    39      9    77%   23-26, 63-66, 70-72
lasagne/layers/noise                    31      0   100%   
lasagne/layers/pool                     79     49    38%   28-29, 58-65, 69-71, 74-85, 102-108, 112-134, 143-144, 147, 150
lasagne/layers/shape                    28     13    54%   21, 24, 31-34, 37-44, 47
lasagne/nonlinearities                  10      0   100%   
lasagne/objectives                      17      0   100%   
lasagne/regularization                   8      4    50%   8-13
lasagne/tests/conftest                   6      1    83%   10
lasagne/tests/layers/conftest            6      0   100%   
lasagne/tests/layers/test_base          90      0   100%   
lasagne/tests/layers/test_conv          43      0   100%   
lasagne/tests/layers/test_dense         55      0   100%   
lasagne/tests/layers/test_input         23      0   100%   
lasagne/tests/layers/test_merge         26      0   100%   
lasagne/tests/layers/test_noise         55      0   100%   
lasagne/tests/test_examples             25      1    96%   36
lasagne/tests/test_init                 64      0   100%   
lasagne/tests/test_nonlinearities        6      0   100%   
lasagne/tests/test_objectives           36      0   100%   
lasagne/theano_extensions/__init__       0      0   100%   
lasagne/theano_extensions/conv         113    104     8%   17-35, 42-56, 63-77, 86-116, 123-170, 177-211
lasagne/theano_extensions/padding       20     16    20%   16-38
lasagne/updates                         59     41    31%   12-18, 22-31, 55-64, 75-84, 96-114
lasagne/utils                           45      9    80%   20-21, 35-36, 48-51, 98, 108
------------------------------------------------------------------
TOTAL                                 1619    515    68%   
```

This is on my laptop (with GPU but no cudnn)

**Comment by dnouri** [comment]
On 01/28/2015 07:22 PM, Sander Dieleman wrote:
> Time for an update now that #114
> <https://github.com/benanne/Lasagne/pull/114> is merged:
> 
> |Name                                 Stmts   Miss  Cover   Missing
> ------------------------------------------------------------------
> lasagne/__init__                         7      0   100%   
> ...
> ------------------------------------------------------------------
> TOTAL                                 1619    515    68%   
> |

It's strangely not picking up the settings from .coveragerc for you,
which tell it to ignore files in lasagne/tests/* for the coverage
report.  Wonder why that would be; it seems it's picked up automatically
for me, despite that I haven't specified the --cov-config option anywhere.

**Comment by benanne** [comment]
Where is .coveragerc supposed to be? It's not in the repository as far as I can tell.

**Comment by dnouri** [comment]
That might be the reason why it's not working.  :-)

**Comment by benanne** [comment]
Now I get:

```
Name                                 Stmts   Miss  Cover   Missing
------------------------------------------------------------------
lasagne/__init__                         7      0   100%   
lasagne/conftest                         3      0   100%   
lasagne/init                            63      3    95%   17, 44, 96
lasagne/layers/__init__                 21      4    81%   13-14, 18-19
lasagne/layers/base                     52      5    90%   163, 211, 216, 262, 289
lasagne/layers/conv                    116     61    47%   22-42, 45-46, 49, 52, 55-64, 69-92, 101, 114, 135-142, 157-164, 167
lasagne/layers/corrmm                   71     18    75%   25, 39, 50, 54-65, 69, 74, 83, 86, 105, 109
lasagne/layers/cuda_convnet            177     64    64%   27, 45, 50, 53, 56, 67, 71-82, 86, 92, 103-104, 117-118, 126, 133, 154, 163, 166, 171, 177, 188-190, 198, 210, 225, 228, 239, 242, 257-276, 279, 282, 285, 288-300
lasagne/layers/dense                    66     29    56%   101-120, 123, 126, 129, 132-146
lasagne/layers/dnn                      90     81    10%   16-137
lasagne/layers/helper                   45      1    98%   58
lasagne/layers/input                    24      2    92%   42-43
lasagne/layers/merge                    39      9    77%   23-26, 63-66, 70-72
lasagne/layers/noise                    31      0   100%   
lasagne/layers/pool                     79     49    38%   28-29, 58-65, 69-71, 74-85, 102-108, 112-134, 143-144, 147, 150
lasagne/layers/shape                    28     13    54%   21, 24, 31-34, 37-44, 47
lasagne/nonlinearities                  10      0   100%   
lasagne/objectives                      17      0   100%   
lasagne/regularization                   8      4    50%   8-13
lasagne/theano_extensions/__init__       0      0   100%   
lasagne/theano_extensions/conv         113    104     8%   17-35, 42-56, 63-77, 86-116, 123-170, 177-211
lasagne/theano_extensions/padding       20     16    20%   16-38
lasagne/updates                         59     41    31%   12-18, 22-31, 55-64, 75-84, 96-114
lasagne/utils                           47     11    77%   20-24, 38-39, 51-54, 101, 111
------------------------------------------------------------------
TOTAL                                 1186    515    57%   
```

Doesn't look too terrible :) I really need to write some tests for those 1D convolution implementations.

**Comment by skaae** [comment]
Whats your opinion about porting torchs tests for updates? 

https://github.com/torch/optim/tree/master/test ?

**Comment by f0k** [comment]
What do they test, though? They just [run the optimizer and print a table of results](https://github.com/torch/optim/blob/master/test/test_sgd.lua), don't they?

**Comment by skaae** [comment]
I think it bascially test that the opitimizer goes downhill/converge? even though they does not have an assert.

**Comment by skaae** [comment]
I think we could use i like this:

```Python
import lasagne
import numpy as np
import theano
lasagne.updates.sgd
def rosenbrock(optimizer):
    x = theano.shared(np.zeros((1,2), dtype='float32'))
    x1 = x[0,0]
    x2 = x[0,1]
    f = 100*(x2-x1**2)**2 + (1-x1)**2
    updates = optimizer(f, [x], learning_rate=1e-3)
    test_grad = theano.function([],f, updates=updates)
    f_ = theano.function([],)
    for i in range(10000):
        res = test_grad()
return [x.get_value(), res]

# assert in test case
print rosenbrock(lasagne.updates.sgd)
```

The minimum is 0 at 1,1.

**Comment by f0k** [comment]
> I think it bascially test that the opitimizer goes downhill/converge?

I think it's just a demonstration to compare how the optimizers perform on a particularly pathological optimization problem. We could put this into the `examples` directory, though.

>  It would just test that all optimizer work and are going downhill.

A better functional test for the optimizers would be to check whether they construct the correct update expressions (this needs to be done separately for each optimizer). We'd actually want to ensure that `sgd` always does vanilla stochastic gradient descent, which is more specific than checking if it goes downhill.

**Comment by skaae** [comment]
I does test for basic programming erors, sign errors etc. Until someone writes specific tests for each update function it is better than no tests :)

I updated the code example

**Comment by benanne** [comment]
> A better functional test for the optimizers would be to check whether they construct the correct update expressions (this needs to be done separately for each optimizer). We'd actually want to ensure that `sgd` always does vanilla stochastic gradient descent, which is more specific than checking if it goes downhill.

Agreed, actually testing whether the optimizers do something sensible seems a little bit outside of the scope of Lasagne. We should just test whether what we have implemented matches what's written in the papers, and let the paper writers deal with showing that their methods actually work :)

**Comment by skaae** [comment]
The purpose of this tests is not to test the performance on that particular problem, but only to test that the optimizers "work" in the sense that they actually go downhill.

I agree that it would be best to test the implementations are as in the papers. But given that it is probably the same person who writes the test and the implementation they might mix things up. I know that happens from personal experience :). In which case it is nice to check that the optimizers are doing something sensible on a "real" problem.  Secondly I think it is better to have these tests than none at all. 

This is only a suggestion, so if you dont agree we can close the discussion here.

**Comment by benanne** [comment]
I haven't made up my mind, I'm just not convinced that we should spend time on this. It's not just about development time, but also the time we'll spend waiting to run the tests. Unless we make this set of tests optional by default (it would be good to do that with the convolution implementation tests as well, I guess).

Overall I think testing the implementations is still a better bet - sure, if the same person writes the tests and the actual implementation, they might make the same mistake twice, but at least it'll be twice as easy to notice :)

**Comment by dnouri** [comment]
> I haven't made up my mind, I'm just not convinced that we should 
> spend time on this. It's not just about development time, but also 
> the time we'll spend waiting to run the tests. Unless we make this 
> set of tests optional by default (it would be good to do that with 
> the convolution implementation tests as well, I guess).
> 
There's always "@pytest.mark.slow" to put around these of course.

**Comment by benanne** [comment]
True. Alright, fair enough. If someone really wants to do this, feel free, as far as I'm concerned :) Although I think right now there are more important things that need to be done first!

**Comment by benanne** [comment]
Now that our API docs are almost complete, one of the last big blockers for our first official release is test coverage. All things considered our test suite is already pretty awesome. I am usually too lazy to test my code, but thanks to a lot of people Lasagne has a ton of tests already, and it's saved me a lot of time and headaches so far. So thank you guys for that!

There's a few gaps still that are probably important to fill up before the first release. The coverage report shown after running `py.test` is really informative in this regard. I will try to do a PR soon to bring the coverage for the various convolution layers as close to 100% as possible.

`layers/pool.py`, `updates.py` and `utils.py` also have relatively low coverage, so PRs to fix this would be welcome too! Don't worry about anything in `theano_extensions` for now, that code is poorly documented and probably has a bunch of bugs as well, so I should spend some time adding tests for that (but I probably won't get around to it soon).

**Comment by skaae** [comment]
I looked at `regularization.py` doesn't that need to be changed to the new tagging system?

**Comment by benanne** [comment]
It needs to be revamped or removed, see #14

**Comment by ebenolson** [comment]
Coverage update - files still needing additional tests are 

- [ ] lasagne/updates
- [ ] lasagne/layers/cuda_convnet
- [ ] lasagne/layers/pool
- [ ] lasagne/layers/helper
- [ ] lasagne/layers/dnn
- [ ] lasagne/layers/conv
- [ ] lasagne/layers/corrmm
- [ ] ~~lasagne/objectives~~
- [ ] ~~lasagne/regularization~~

objectives and regularization will likely change before release, so not a priority at the moment.

@benanne is handling the conv-related (cuda_convnet, dnn, conv, corrmm), leaving pool, helper and updates.

**Comment by benanne** [comment]
Now that we are approaching 100%, does anybody object to adding a coverage badge to our README file?

I'm not sure if it's possible to get a badge that's based on pytest-cov's output on Travis - I doubt it, to be honest. But we can also have a look at coveralls.io, which seems to have a nice interface for checking how commits / PRs affect coverage.

**Comment by f0k** [comment]
> I'm not sure if it's possible to get a badge that's based on pytest-cov's output on Travis - I doubt it, to be honest.

It seems it should be possible somehow: https://github.com/z4r/python-coveralls#travisyml
I don't have much of a clue about the travis/coveralls/pytest ecosystem, though.

**Comment by benanne** [comment]
With #267 merged, we just need to deal with `objectives` and `regularization`. These submodules still need to be reworked and some new tests will probably have to be written for them, but I think it's probably okay to close this issue now. Or we can leave it open for a bit longer as a reminder to write enough tests for these submodules as well. Either is fine by me.

**Comment by f0k** [comment]
> I think it's probably okay to close this issue now

Yes, do it! :) Good job, Eben and Sander! :+1:
